crates/miden-tx/src/executor/exec_host.rs

```diff
@@ -212,6 +213,69 @@ where
         Ok(Vec::new())
     }
 
+    /// Handles a request to an asset witness by querying the data store for a merkle path.
+    ///
+    /// ## Native Account
+    ///
+    /// For the native account we always request witnesses for the initial vault root, because the
+    /// data store only has the state of the account vault at the beginning of the transaction.
+    /// Since the vault root can change as the transaction progresses, this means the witnesses
+    /// may become _partially_ or fully outdated. To see why they can only be _partially_ outdated,
+    /// consider the following example:
+    ///
+    /// ```text
+    ///      A               A'
+    ///     / \             /  \
+    ///    B   C    ->    B'    C
+    ///   / \  / \       /  \  / \
+    ///  D  E F   G     D   E' F  G
+    /// ```
+    ///
+    /// Leaf E was updated to E', in turn updating nodes B and A. If we now request the merkle path
+    /// to G against root A (the initial vault root), we'll get nodes F and B. F is a node in the
+    /// updated tree, while B is not. We insert both into the merkle store anyway. Now, if the
+    /// transaction attempts to verify the merkle path to G, it can do so because F and B' are in
+    /// the merkle store. Note that B' is in the store because the transaction inserted it into the
+    /// merkle store as part of updating E, not because we inserted it. B is present in the store,
+    /// but is simply ignored for the purpose of verifying G's inclusion.
+    async fn on_account_vault_asset_witness_requested(
+        &self,
+        current_account_id: AccountId,
+        _vault_root: Word,
+        asset: Asset,
+    ) -> Result<Vec<AdviceMutation>, TransactionKernelError> {
+        // For now, we only support getting witnesses for the native account, so return early if the
+        // requested account is not the native one.
+        if current_account_id != self.base_host.native_account_header().id() {
+            return Ok(Vec::new());
+        }
+
+        let vault_root = self.base_host.native_account_header().vault_root();
```

Comment:
I'm not sure I fully understand why `vault_root` is passed to the function and then ignored. The explanation about always fetching data for the initial vault root makes sense, but in what scenarios would this change? 

In the meantime, should there be an `assert_eq!()` for checking that the passed in root is the same as this one or does this not make sense?